File: packages/engine/src/node-executors/mcp.ts
/** External MCP client node executor and its timeline bridge. */

import { getOrgConfigSnapshot } from "@janusly/data";

import {
  executeMcpTool,
  readMcpClientWritesEnabled,
  resolveMcpClientRateLimitPerMin,
  resolveStdioSandboxConfig,
} from "../mcp-tool-executor";
import { appendEvent } from "../persistence";
import { mapInput } from "../template";
import type { NodeExecutorMap } from "./types";

export const mcpNodeExecutors = {
  // mcp_tool node — invokes a registered external MCP server's tool
  // through `executeMcpTool`. The executor enforces:
  //   1. multi-tenant scope (org-scoped repo lookups);
  //   2. connection + tool enabled flags;
  //   3. write-side dry-run skip (sandbox replay);
  //   4. two-flag write consent (process env + tenant config);
  //   5. env-ref resolution (generic error on miss; never echoes env var name);
  //   6. per-tool rate-limit via `getEngineRateLimiter()`;
  //   7. timeout + error containment.
  // The envelope is `{ ok, output?, error?, latencyMs }` — the
  // executor NEVER throws on runtime failures. Failure cases throw
  // here so the run's existing retry / DLQ machinery applies (matches
  // the `http` node contract).
  mcp_tool: async (ctx) => {
    const { connectionAlias, toolName, input, timeoutMs } = ctx.config;
    if (typeof connectionAlias !== "string" || !connectionAlias) {
      throw new Error("mcp_tool requires config.connectionAlias");
    }
    if (typeof toolName !== "string" || !toolName) {
      throw new Error("mcp_tool requires config.toolName");
    }

    const mappedInput = input && typeof input === "object" && !Array.isArray(input)
      ? mapInput(input as Record<string, unknown>, { context: ctx.context, inputs: ctx.config })
      : {};
    const orgConfig = await getOrgConfigSnapshot(ctx.orgId);

    await appendEvent(ctx.runId, ctx.nodeId, "mcp_tool.started", {
      connectionAlias,
      toolName,
    });
    const envelope = await executeMcpTool({
      orgId: ctx.orgId,
      connectionAlias,
      toolName,
      input: mappedInput as Record<string, unknown>,
      timeoutMs,
      dryRun: ctx.dryRun,
      runId: ctx.runId,
      nodeId: ctx.nodeId,
      workflowId: ctx.workflowId ?? undefined,
      writeConsentProcess: readMcpClientWritesEnabled(),
      writeConsentTenant: orgConfig.mcp.clientWriteConsent,
      rateLimitPerMin: resolveMcpClientRateLimitPerMin(orgConfig.mcp.clientRateLimitPerMin),
      stdioSandbox: resolveStdioSandboxConfig(orgConfig.mcp),
      onAudit: async ({ ok, error, latencyMs, writeSide, sandboxFailureCode, capturedStderrTail }) => {
        // Always emit the per-call timeline event so the run timeline
        // shows the outcome. Sandbox-driven kills get an extra
        // `mcp.sandbox.terminated`-shaped event (in addition to the
        // completed event) carrying the typed reason + redacted tail so
        // operators see WHY the call ended without re-running the call.
        await appendEvent(ctx.runId, ctx.nodeId, "mcp_tool.completed", {
          connectionAlias,
          toolName,
          ok,
          error,
          latencyMs,
          writeSide,
          ...(sandboxFailureCode ? { sandboxFailureCode } : {}),
          ...(capturedStderrTail ? { stderrTail: capturedStderrTail } : {}),
        });
        if (sandboxFailureCode) {
          await appendEvent(ctx.runId, ctx.nodeId, "mcp.sandbox.terminated", {
            connectionAlias,
            toolName,
            reason: sandboxFailureCode,
            stderrTail: capturedStderrTail ?? null,
            writeSide,
          });
        }
      },
    });
    if (!envelope.ok) {
      throw new Error(`mcp_tool failed: ${envelope.error ?? "unknown"}`);
    }
    return { status: "completed", output: envelope.output ?? {} };
  },
} satisfies Pick<NodeExecutorMap, "mcp_tool">;

File: internal/executors/mcp.go
// The `mcp_tool` node executor — the thin, dependency-free half of the
// MCP client. All heavy lifting (connection/descriptor lookup, consent,
// sandbox, transports) lives behind the injected McpDeps.Execute closure
// (built by the engine dispatcher from internal/mcpclient), so this
// package never imports the MCP SDK or the store. The chokepoint NEVER
// throws — it returns a typed envelope on every path — and this executor
// converts `ok:false` into an error so the run's existing retry / DLQ
// machinery applies (same contract as the http node).
package executors

import (
	"context"
	"fmt"
)

// McpCall is one tool invocation as the node config expresses it.
type McpCall struct {
	ConnectionAlias string
	ToolName        string
	Input           map[string]any
	TimeoutMs       float64
}

// McpEnvelope is the normalized per-call outcome, implements the
// contract's McpToolEnvelope.
type McpEnvelope struct {
	OK              bool
	Error           string
	Output          map[string]any
	LatencyMs       int64
	ConnectionAlias string
	ToolName        string
	Transport       string
	WriteSide       bool
	// SandboxFailureCode is set only for stdio terminations driven by the
	// sandbox layer (allowlist / lifetime / stderr cap); empty otherwise.
	SandboxFailureCode string
	// StderrTail is the last 1024 chars of the redacted stdio stderr
	// capture; empty for URL transports.
	StderrTail string
}

// McpDeps is the dispatcher-built seam carrying the real client.
type McpDeps struct {
	Execute func(ctx context.Context, call McpCall) McpEnvelope
}

// NewMcpToolExecutor builds the mcp_tool node over the injected seam.
func NewMcpToolExecutor() Func {
	return func(ctx context.Context, in Input) (any, error) {
		alias, _ := in.Config["connectionAlias"].(string)
		if alias == "" {
			return nil, fmt.Errorf("mcp_tool requires config.connectionAlias")
		}
		toolName, _ := in.Config["toolName"].(string)
		if toolName == "" {
			return nil, fmt.Errorf("mcp_tool requires config.toolName")
		}
		if in.Mcp == nil || in.Mcp.Execute == nil {
			return nil, fmt.Errorf("mcp_tool executor requires engine context")
		}
		input, _ := in.Config["input"].(map[string]any)
		if input == nil {
			input = map[string]any{}
		}
		timeoutMs, _ := in.Config["timeoutMs"].(float64)

		emit := func(eventType string, payload map[string]any) {
			if in.Emit != nil {
				in.Emit(eventType, payload)
			}
		}
		emit("mcp_tool.started", map[string]any{
			"connectionAlias": alias, "toolName": toolName,
		})
		envelope := in.Mcp.Execute(ctx, McpCall{
			ConnectionAlias: alias, ToolName: toolName,
			Input: input, TimeoutMs: timeoutMs,
		})
		completed := map[string]any{
			"connectionAlias": alias, "toolName": toolName,
			"ok": envelope.OK, "latencyMs": envelope.LatencyMs,
			"writeSide": envelope.WriteSide,
		}
		if envelope.Error != "" {
			completed["error"] = envelope.Error
		}
		if envelope.SandboxFailureCode != "" {
			completed["sandboxFailureCode"] = envelope.SandboxFailureCode
		}
		if envelope.StderrTail != "" {
			completed["stderrTail"] = envelope.StderrTail
		}
		emit("mcp_tool.completed", completed)
		if envelope.SandboxFailureCode != "" {
			var tail any
			if envelope.StderrTail != "" {
				tail = envelope.StderrTail
			}
			emit("mcp.sandbox.terminated", map[string]any{
				"connectionAlias": alias, "toolName": toolName,
				"reason": envelope.SandboxFailureCode, "stderrTail": tail,
				"writeSide": envelope.WriteSide,
			})
		}
		if !envelope.OK {
			errText := envelope.Error
			if errText == "" {
				errText = "unknown"
			}
			return nil, fmt.Errorf("mcp_tool failed: %s", errText)
		}
		output := envelope.Output
		if output == nil {
			output = map[string]any{}
		}
		return map[string]any{"status": "completed", "output": output}, nil
	}
}
